File: user-service/src/main/java/com/athena/userservice/service/PasswordPolicyService.java
package com.athena.userservice.service;

import com.athena.userservice.model.PasswordPolicy;
import com.athena.userservice.repository.PasswordPolicyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
public class PasswordPolicyService {

    private final PasswordPolicyRepository policyRepository;

    public void validatePassword(String password) {
        PasswordPolicy policy = policyRepository.findAll().stream().findFirst().orElse(null);
        if (policy == null) return;

        if (password.length() < policy.getMinLength()) {
            throw new RuntimeException("Password must be at least " + policy.getMinLength() + " characters long");
        }
        if (policy.isRequireUppercase() && !Pattern.compile("[A-Z]").matcher(password).find()) {
            throw new RuntimeException("Password must contain at least one uppercase letter");
        }
        if (policy.isRequireLowercase() && !Pattern.compile("[a-z]").matcher(password).find()) {
            throw new RuntimeException("Password must contain at least one lowercase letter");
        }
        if (policy.isRequireNumbers() && !Pattern.compile("[0-9]").matcher(password).find()) {
            throw new RuntimeException("Password must contain at least one number");
        }
        if (policy.isRequireSpecialChars()) {
            String specialChars = policy.getSpecialChars() != null && !policy.getSpecialChars().isEmpty()
                    ? Pattern.quote(policy.getSpecialChars())
                    : "!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>\\/?";
            if (!Pattern.compile("[" + specialChars + "]").matcher(password).find()) {
                throw new RuntimeException("Password must contain at least one special character");
            }
        }
    }

    public PasswordPolicy getCurrentPolicy() {
        return policyRepository.findAll().stream().findFirst()
                .orElseThrow(() -> new RuntimeException("No password policy configured"));
    }

    public PasswordPolicy updatePolicy(PasswordPolicy newPolicy) {
        PasswordPolicy existing = policyRepository.findAll().stream().findFirst().orElse(new PasswordPolicy());
        existing.setMinLength(newPolicy.getMinLength());
        existing.setRequireUppercase(newPolicy.isRequireUppercase());
        existing.setRequireLowercase(newPolicy.isRequireLowercase());
        existing.setRequireNumbers(newPolicy.isRequireNumbers());
        existing.setRequireSpecialChars(newPolicy.isRequireSpecialChars());
        existing.setExpirationDays(newPolicy.getExpirationDays());
        existing.setSpecialChars(newPolicy.getSpecialChars());
        return policyRepository.save(existing);
    }
}

File: user-service/internal/service/password_policy_service.go
package service

import (
	"fmt"
	"regexp"

	"github.com/athena/user-service/internal/model"
	"github.com/athena/user-service/internal/repository"
)

type PasswordPolicyService struct {
	repo *repository.PasswordPolicyRepository
}

func NewPasswordPolicyService(repo *repository.PasswordPolicyRepository) *PasswordPolicyService {
	return &PasswordPolicyService{repo: repo}
}

// ValidatePassword validates a password against the current policy.
// Returns nil if no policy is configured.
func (s *PasswordPolicyService) ValidatePassword(password string) error {
	policy, err := s.repo.FindFirst()
	if err != nil {
		// No policy configured — allow any password
		return nil
	}

	if len(password) < policy.MinLength {
		return fmt.Errorf("Password must be at least %d characters long", policy.MinLength)
	}
	if policy.RequireUppercase {
		if matched, _ := regexp.MatchString(`[A-Z]`, password); !matched {
			return fmt.Errorf("Password must contain at least one uppercase letter")
		}
	}
	if policy.RequireLowercase {
		if matched, _ := regexp.MatchString(`[a-z]`, password); !matched {
			return fmt.Errorf("Password must contain at least one lowercase letter")
		}
	}
	if policy.RequireNumbers {
		if matched, _ := regexp.MatchString(`[0-9]`, password); !matched {
			return fmt.Errorf("Password must contain at least one number")
		}
	}
	if policy.RequireSpecialChars {
		specialPattern := `[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>/?]`
		if policy.SpecialChars != "" {
			specialPattern = "[" + regexp.QuoteMeta(policy.SpecialChars) + "]"
		}
		if matched, _ := regexp.MatchString(specialPattern, password); !matched {
			return fmt.Errorf("Password must contain at least one special character")
		}
	}
	return nil
}

// GetCurrentPolicy returns the current password policy.
func (s *PasswordPolicyService) GetCurrentPolicy() (*model.PasswordPolicy, error) {
	policy, err := s.repo.FindFirst()
	if err != nil {
		return nil, fmt.Errorf("No password policy configured")
	}
	return policy, nil
}

// UpdatePolicy updates the current password policy, or creates one if none exists.
func (s *PasswordPolicyService) UpdatePolicy(newPolicy *model.PasswordPolicy) (*model.PasswordPolicy, error) {
	existing, err := s.repo.FindFirst()
	if err != nil {
		// No existing policy — create one
		existing = &model.PasswordPolicy{}
	}
	existing.MinLength = newPolicy.MinLength
	existing.RequireUppercase = newPolicy.RequireUppercase
	existing.RequireLowercase = newPolicy.RequireLowercase
	existing.RequireNumbers = newPolicy.RequireNumbers
	existing.RequireSpecialChars = newPolicy.RequireSpecialChars
	existing.ExpirationDays = newPolicy.ExpirationDays
	existing.SpecialChars = newPolicy.SpecialChars

	if err := s.repo.Save(existing); err != nil {
		return nil, err
	}
	return existing, nil
}
